kaldb/src/main/java/com/slack/kaldb/metadata/core/CuratorBuilder.java

```diff
@@ -70,6 +71,7 @@ public static AsyncCuratorFramework build(
                   && curatorEvent.getWatchedEvent().getState()
                       == Watcher.Event.KeeperState.Expired) {
                 LOG.warn("The ZK session has expired {}.", curatorEvent);
+                new RuntimeHalterImpl().handleFatal(new Throwable("ZK session expired."));
```

Comment:
Correct, this is a fix that is not directly related to this PR but was exposed after deploying this PR to dev. The autoscaling kept growing when it wasn't expected to, and troubleshooting that issue revealed that cache slots were disappearing once ZK reconnected. This appears to have settled down the autoscaling to be more in line with what we were expecting, but we'll need some longer term data to ensure there's not additional issues causing the slots to disappear.

This would be an important fix to keep (and corresponding tests updates), regardless of the rest of this PR.